File: C++/代码随想录/Binary_Tree/Modification and construction of binary Search tree/450. Delete Node in a BST.py
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def deleteNode(self, root: Optional[TreeNode], key: int) -> Optional[TreeNode]:
        if not root: return root

        if root.val > key:
            root.left = self.deleteNode(root.left, key)
        elif root.val < key:
            root.right = self.deleteNode(root.right, key)
        else:
            if not root.left: return root.right
            if not root.right: return root.left

            if root.left and root.right:
                temp = root.right
                while temp.left: temp = temp.left
                root.val = temp.val
                root.right = self.deleteNode(root.right, root.val)
        
        return root  

# Test case
root = TreeNode(5, TreeNode(3, TreeNode(2), TreeNode(4)), TreeNode(6, None, TreeNode(7)))
sol = Solution()
updated_tree = sol.deleteNode(root, 3)

# 测试用例
def inorder_traversal(root):
    return inorder_traversal(root.left) + [root.val] + inorder_traversal(root.right) if root else []
print(inorder_traversal(updated_tree)) # [2, 4, 5, 6, 7]

File: C++/代码随想录/Binary_Tree/Modification and construction of binary Search tree/450. Delete Node in a BST.cpp
// 450. Delete Node in a BST
// https://leetcode.com/problems/delete-node-in-a-bst/
// clang++ 450.\ Delete\ Node\ in\ a\ BST.cpp -std=c++17
#include "../../printUtils.hpp"

template<typename T>
class Solution {
public:
    // Find the minimum value in the right subtree
    int findMin(TreeNode<T>* root) {
        while (root->left) {
            root = root->left;
        }
        return root->val;
    }
    // Delete a node from a BST and return the new root
    TreeNode<T>* deleteNode(TreeNode<T>* root, int key) {
        if (!root) return nullptr;
        
        // If the key is smaller than the root, delete from the left subtree
        if (key < root->val) {
            root->left = deleteNode(root->left, key);
        }
        // If the key is larger than the root, delete from the right subtree
        else if (key > root->val) {
            root->right = deleteNode(root->right, key);
        }
        // If the key is equal to the root, delete the root
        else {
            // Case 1: no child
            if (!root->left && !root->right) {
                delete root;
                return nullptr;
            }
            // Case 2: one child
            else if (!root->left || !root->right) {
                TreeNode<T>* temp = root->left ? root->left : root->right;
                delete root;
                return temp;
            }
            // Case 3: two children
            else {
                int successor = findMin(root->right);
                root->val = successor;
                root->right = deleteNode(root->right, successor);
            }
        }
        return root;
    }
};

int main() {
    TreeNode<int>* root = buildTree({5, 3, 6, 2, 4, -1, 7}, -1);
    int key = 3;
    TreeNode<int>* result = Solution<int>().deleteNode(root, key);
    std::cout << "The result is: " << std::endl;
    printArray(levelOrder(result));

    TreeNode<int>* root2 = buildTree({5, 3, 6, 2, 4, -1, 7}, -1);
    int key2 = 0;
    TreeNode<int>* result2 = Solution<int>().deleteNode(root2, key2);
    std::cout << "The result is: " << std::endl;
    printArray(levelOrder(result2));

    return 0;
}

